### PR TITLE
Warns on read-only Numpy  array->tensor conversion

### PR DESCRIPTION
**Pull request by mruberry**
Addresses https://github.com/pytorch/pytorch/issues/5442.

Per title (and see issue). A test is added to test_torch.py to verify the behavior.

Update (with new behavior): 

NumPy arrays can be non-writeable (read-only). When converting a NumPy array to a Torch tensor the storage is shared, but the tensor is always writable (PyTorch doesn't have a read-only tensor). Thus, when a non-writeable NumPy array is converted to a PyTorch tensor it can be written to.

In the past, PyTorch would silently copy non-writeable NumPy arrays and then convert those copies into tensors. This behavior violates the from_numpy contract, however, which promises that the tensor and the array share memory.

This PR adds a warning message when a non-writeable NumPy array is converted into a Torch tensor. This will not break any networks, but will make end users aware of the behavior. They can work-around the warning message by marking their NumPy arrays as writeable.

### CODE REVIEW

**Comment by mruberry** [comment]
@ailzhang If this goes in it will break an XLA test that creates a tensor from a read-only Numpy array.

**Comment by ailzhang** [comment]
FYI the xla failure here is caused by use of `torch.from_numpy` in torchvision. So we'll probably need to fix that if we want to throw an error for this. 

```
======================================================================
ERROR: test_accurracy (__main__.TrainMnist)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_train_mnist.py", line 186, in test_accurracy
    self.assertGreaterEqual(train_mnist(), FLAGS.target_accuracy)
  File "test_train_mnist.py", line 74, in train_mnist
    transforms.Normalize((0.1307,), (0.3081,))]))
  File "/private/home/ailzhang/miniconda/envs/teng/lib/python3.6/site-packages/torchvision-0.6.0a0+0156d58-py3.6-linux-x86_64.egg/torchvision/datasets/mnist.py", line 71, in __init__
    self.download()
  File "/private/home/ailzhang/miniconda/envs/teng/lib/python3.6/site-packages/torchvision-0.6.0a0+0156d58-py3.6-linux-x86_64.egg/torchvision/datasets/mnist.py", line 144, in download
    read_image_file(os.path.join(self.raw_folder, 'train-images-idx3-ubyte')),
  File "/private/home/ailzhang/miniconda/envs/teng/lib/python3.6/site-packages/torchvision-0.6.0a0+0156d58-py3.6-linux-x86_64.egg/torchvision/datasets/mnist.py", line 483, in read_image_file
    x = read_sn3_pascalvincent_tensor(f, strict=False)
  File "/private/home/ailzhang/miniconda/envs/teng/lib/python3.6/site-packages/torchvision-0.6.0a0+0156d58-py3.6-linux-x86_64.egg/torchvision/datasets/mnist.py", line 470, in read_sn3_pascalvincent_tensor
    return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)
TypeError: The given Numpy array is not writeable, and PyTorch does not support non-writeable tensors. Copy the array or make it writeable.

----------------------------------------------------------------------
Ran 1 test in 2.548s
```

**Comment by mruberry** [comment]
Thanks for the follow-up, @ailzhang. I think that suggests going with a warning if this happens is the right approach, at least for now. I'll make the changes soon.

File: torch/csrc/utils/tensor_numpy.cpp [hunk 1]
```diff
@@ -129,8 +129,17 @@ at::Tensor tensor_from_numpy(PyObject* obj) {
   if (!PyArray_Check(obj)) {
     throw TypeError("expected np.ndarray (got %s)", Py_TYPE(obj)->tp_name);
   }
-
   auto array = (PyArrayObject*)obj;
+
+  if (!PyArray_ISWRITEABLE(array)) {
+    TORCH_WARN(
```
Comment (gchanan) [inline torch/csrc/utils/tensor_numpy.cpp]: perhaps we should TORCH_WARN_ONCE?  I'm not really sure I can think of a case where we wouldn't want to warn once...users really don't like getting spammed.

File: test/test_torch.py [hunk 2]
```diff
@@ -3293,6 +3299,8 @@ def _generate_noncontiguous_input():
 
             out = np.broadcast_to(np.random.random((1, 4)),
                                   initial_shape)
+            # Note: non-writeable NumPy arrays will warn if converted to tensors
+            out.setflags(write=True)
```
Comment (gchanan) [inline test/test_torch.py]: just out of curiosity -- this is necessary?  What makes NumPy arrays (by default?) non-writeable?

**Review by gchanan — APPROVED** [review]

File: test/test_torch.py (hunk 2 above)
Comment (mruberry) [reply test/test_torch.py]: NumPy arrays are not, by default, non-writeable, but many NumPy operations produce non-writeable arrays. For example, [numpy.diagonal](https://numpy.org/doc/1.18/reference/generated/numpy.diagonal.html#numpy.diagonal) returns a read-only view. And in this case [broadcast_to](https://numpy.org/doc/1.18/reference/generated/numpy.broadcast_to.html?highlight=broadcast_to#numpy.broadcast_to) returns a non-writeable view.

Comment (gchanan) [reply test/test_torch.py]: ah that's neat.  Although I don't know why we even use a NumPy array here...

File: torch/csrc/utils/tensor_numpy.cpp (hunk 1 above)
Comment (mruberry) [reply torch/csrc/utils/tensor_numpy.cpp]: Updated.

**Pull request closed without merging**